### ISSUE TITLE
Yaml anchors and aliases across actions are no longer supported in porter.yaml

### ISSUE THREAD

**Issue by jasmdk** [getporter/porter#2281]
🐱 I have added a comment that describes our desired solution: https://github.com/getporter/porter/issues/2281#issuecomment-1205416159 (carolynvs)

## Describe the bug
Although it has not been described as a feature, but we have previously relied on YAML anchors and aliases (https://www.educative.io/blog/advanced-yaml-syntax-cheatsheet#anchors) to avoid duplicating steps in install and upgrade which is otherwise the same, but from v1.0.0-beta.1 to v1.0.0-beta.2 this feature is no longer supported - maybe in an attempt to optimize and only look at the action to be invoked.
Any chance this can be reverted or do we have to duplicate all similar actions?

## To Reproduce
1. Create a step in one action that references a step in another - here an example using stateless custom actions:
```
mystateless:
  - exec: &statelessAlias
      description: "Invoke my stateless action"
      command: ./helpers.sh
      arguments:
        - stateless

mystateless-other:
  - exec: *statelessAlias

customActions:
  mystateless:
    description: "This is a stateless action"
    modifies: false
    stateless: true

  mystateless-other:
    description: "This is a another stateless action"
    modifies: false
    stateless: true
```
3. Run `porter invoke --action mystateless`
4. The action executes without issues
5. Run `porter invoke --action mystateless-other`
7. See error

## Expected behavior
Both actions need the same step and to avoid a lot of duplicate YAML code, an anchor can be set on one YAML entity and referenced in another. An extremely useful case is with helm mixin and airgapped scenario where all image values must be set explicitly using the porter calculated image digests. If you then want to install the Helm release during "install" but allow changes to the parameters while invoking "upgrade" then all the image settings must be duplicated and kept in sync.

## Porter Command and Output
```
$ porter invoke --action mystateless
executing mystateless action from porter-hello (installation: synergy/porter-hello)
Invoke my stateless action
Doing stateless stuff
execution completed successfully!


$ porter invoke --action mystateless-other
executing mystateless-other action from porter-hello (installation: synergy/porter-hello)
invalid step yaml after rendering template
{{=${ }=}}
exec: *statelessAlias
: yaml: unknown anchor 'statelessAlias' referenced
1 error occurred:
        * unable to resolve step: invalid step yaml after rendering template
{{=${ }=}}
exec: *statelessAlias
: yaml: unknown anchor 'statelessAlias' referenced


1 error occurred:
        * unable to resolve step: invalid step yaml after rendering template
{{=${ }=}}
exec: *statelessAlias
: yaml: unknown anchor 'statelessAlias' referenced
```

## Version
porter v1.0.0-beta.2 (b2f23785)

**Comment by jarnfast** [comment]
Possibly related to this commit: [https://github.com/getporter/porter/commit/a9f1d29e0fc9d9c2fd37897a126bd6ce9e573743](https://github.com/getporter/porter/commit/a9f1d29e0fc9d9c2fd37897a126bd6ce9e573743)

**Comment by carolynvs** [comment]
Yeah we never said we supported that for a reason. 😀 You are correct that the mentioned commit (where I fixed some long-standing bugs in how we parse and evaluate templates) is what caused anchors to stop working.

https://github.com/getporter/porter/blob/afcd5deaf6f5cef22f84b85411896d72c25249e8/pkg/runtime/runtime_manifest.go#L431-L442

Maybe if we rendered the entire porter.yaml instead of just the single step's YAML node, and then use the stepPath to identify the fully rendered node's YAML, we can get back anchor functionality.

**Comment by carolynvs** [comment]
One thing I realized is that by rendering the entire porter.yaml as a template, we won't be able to validate that the template didn't use a variable that don't exist (since later steps after the current step may reference outputs that don't exist yet). If you decide to try to change how we render the step template, that is something that we don't want to lose. 

One way to address that would be to move the check for unused template variables from runtime to buildtime. We already have similar issues where we want to validate the template at build time, like #1018.

**Comment by jarnfast** [comment]
Unfortunately I have no knowledge of the YAML modules used and thus no idea whether they support normalizing (or as yq named it - 'exploding') the YAML in a way that could be used immediately after reading the file into `m.manifestYAML`.

The new behavior is really messing with us - but I guess that's the risk of using undocumented features 😜 

@jasmdk  Short term I believe we can work around this by normalizing the porter.yaml in our build pipeline, e.g. prior to `porter build`:
```
yq 'explode(.)' ourporter.yaml > porter.yaml
porter build .. 
porter publish ..
```
.. but that said, it would be a really nice feature to re-introduce in porter 😄

**Comment by carolynvs** [comment]
Sorry I broke that for you, I really didn't have a clue that anyone was successfully using anchors! Hopefully we can get this fixed again.

**Comment by carolynvs** [comment]
We need to do two things to restore YAML anchor functionality:

- [ ] [Render the entire manifest instead of just the yaml snippet](https://github.com/getporter/porter/blob/f7aa99dc8cfc0ac48cb019df45ac39e987135c39/pkg/runtime/runtime_manifest.go#L766) associated with the current step that we want to run. This should allow our yaml library to expand out any anchors so that the step contains ALL of the desired logic.
- [ ] [Disable checking that all variables are defined](https://github.com/getporter/porter/blob/f7aa99dc8cfc0ac48cb019df45ac39e987135c39/pkg/runtime/runtime_manifest.go#L441) when we render the yaml template for a step (because we are rendering the entire manifest and later steps may be using variables that have not been generated YET).
- [ ] Add a check to `porter lint` in our [pkg/linter](https://github.com/getporter/porter/blob/release/v1/pkg/linter/linter.go) package that all template variables used are capable of being resolved when the bundle is run. So all parameters/credentials should be defined, no typos on other variables like `bundle.name`, any outputs used, will have been generated by the time the step is run (hard).

Here is a basic plan for determining if an output variable will be resolvable at runtime:

1. Expand the YAML (to get rid of anchors)
2. Walk through each step of each action, accumulate the names of outputs generated by that step and check that any outputs used by the step have been seen already in the previous steps (look in a map of outputs we have seen before).

```
# porter install --param myname=aaron
parameters:
- name: myname

install:
- exec:
     outputs:
      - name: age
- exec:
     arguments: 
     - ${ bundle.parameters.myname }
     - ${ bundle.outputs.age }
```
